### PR TITLE
[Metricbeat] gcp: add GKE metricset

### PR DESCRIPTION
**Pull request by endorama**
## What does this PR do?

Add a lightweight module to pull GCP GKE metrics.

Metrics gathered are all GA metrics as of today available for Kubernetes on GCP.

GCP Docs: https://cloud.google.com/monitoring/api/metrics_kubernetes

## Why is it important?

Implement GKE integration as requested in https://github.com/elastic/integrations/issues/816

## Dates

- Development Start date: `Tue, Aug 17 2021`
- Feature Freeze date: `Wed, Aug 18 2021`
- Release date: `Tue, Sep 21 2021`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

You can build and run the `metricbeat` binary from this PR to collect GKE metrics. You may use this configuration for `gcp.yml` module:

```yaml
- module: gcp
  metricsets:
    - gke
  zone: "us-"
  project_id: "<fill>"
  credentials_file_path: "<fill>"
  exclude_labels: false
  period: 1m
```

## Related issues

- Closes https://github.com/elastic/integrations/issues/816
- Requires https://github.com/elastic/beats/pull/26960
- Requires https://github.com/elastic/beats/pull/26973
- Requires https://github.com/elastic/beats/pull/27286
- Relates https://github.com/elastic/beats/issues/27231
	**NOTE**: due to this issue the fields do not align with Beats naming convention or ECS, to keep the behaviour across Metricsets consistent. Once this is addressed I'll tackle migrating all Metricsets to the correct format.

## Use cases

```
Feature: GKE metric collection

  GKE is the dedicated Kubernetes distribution from GCP. It exposes a set of dedicated 
  metrics useful for Kubernetes operators and developers.

  Scenario: Collect GKE specific metrics
	Given that gke metricset collection has been enabled and configured
    When running metricbeat
	Then I receive GKE specific metrics in Elasticsearch
```

## Screenshots

Screenshot 2021-08-11 at 12-14-55  Metricbeat GCP  GKE Overview 2nd - Elastic

## Logs

# Metricbeat Module / Dataset release checklist

This checklist is intended for Devs which create or update a module to make sure modules are consistent.

## Modules

~~For a metricset to go GA, the following criterias should be met:~~

`gcp` module is currently in `beta`.

## Metricbeat module

* [ ] Example `data.json` exists and an automated way to generate it exists (`go test -data`)
* [ ] Test environment in Docker exist for integration tests

### CODE REVIEW

File: x-pack/metricbeat/module/gcp/gke/manifest.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+default: true
+input:
+  module: gcp
+  metricset: metrics
+  defaults:
+    metrics:
+    - service: gke
+      metric_types:
+      - "container/cpu/core_usage_time"
```
Comment (mtojek) [inline x-pack/metricbeat/module/gcp/gke/manifest.yml:9]: To review available metrics, you can sync with @sorantis .

File: x-pack/metricbeat/module/gcp/gke/_meta/fields.yml [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        description: Local ephemeral storage usage in bytes. Sampled every 60 seconds.
+    - name: memory
+      type: group
+      fields:
+      - name: limit_bytes.value
+        type: long
+        description: Memory limit of the container in bytes. Sampled every 60 seconds.
+      - name: limit_utilization.value
+        type: double
+        description: The fraction of the memory limit that is currently in use on
```
Comment (mtojek) [inline x-pack/metricbeat/module/gcp/gke/_meta/fields.yml:53]: I think we used `pct` for percentages. See:

```
  - name: memory.broker.pct
    description: The percentage of the memory limit used.
    type: scaled_float
    format: percent
```

Sometimes you need an additional Javascript processor to convert to the right format:

```
processors:
  - script:
      lang: javascript
      source: >
        function process(event) {
          var broker_memory_broker_pct = event.Get("activemq.broker.memory.broker.pct")
          if (broker_memory_broker_pct != null) {
            event.Put("activemq.broker.memory.broker.pct", broker_memory_broker_pct / 100.0)
          }

          var broker_memory_temp_pct = event.Get("activemq.broker.memory.temp.pct")
          if (broker_memory_temp_pct != null) {
            event.Put("activemq.broker.memory.temp.pct", broker_memory_temp_pct / 100.0)
          }

          var broker_memory_store_pct = event.Get("activemq.broker.memory.store.pct")
          if (broker_memory_store_pct != null) {
            event.Put("activemq.broker.memory.store.pct", broker_memory_store_pct / 100.0)
          }
        }
```

(based on the ActiveMQ module.

File: x-pack/metricbeat/module/gcp/gke/_meta/fields.yml [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      - name: request_utilization.value
+        type: double
+        description: The fraction of the requested memory that is currently in use
+          on the instance. This value can be greater than 1 as usage can exceed the
+          request. Sampled every 60 seconds. After sampling, data is not visible for
+          up to 240 seconds.
+      - name: used_bytes.value
+        type: long
+        description: Memory usage in bytes. Sampled every 60 seconds.
+    - name: ''
```
Comment (mtojek) [inline x-pack/metricbeat/module/gcp/gke/_meta/fields.yml:73]: Hm.. empty name?

File: x-pack/metricbeat/module/gcp/gke/_meta/fields.yml [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        type: long
+        description: Total ephemeral storage bytes on the node. Sampled every 60 seconds.
+      - name: used_bytes.value
+        type: long
+        description: Local ephemeral storage bytes used by the node. Sampled every
+          60 seconds.
+    - name: memory
+      type: group
+      fields:
+      - name: allocatable_bytes.value
```
Comment (mtojek) [inline x-pack/metricbeat/module/gcp/gke/_meta/fields.yml:131]: Try to follow the Beats naming convention: https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html

**Review by mtojek — COMMENTED** [review]
This is a good start, keep pushing the missing items. You can copy the TODO list from the template: https://github.com/elastic/beats/issues/new?assignees=&labels=&template=module-checklist.md

I think that the next step is to make it working and passing lint.

File: x-pack/metricbeat/module/gcp/gke/_meta/fields.yml (hunk 4 above)
Comment (endorama) [reply x-pack/metricbeat/module/gcp/gke/_meta/fields.yml:131]: Should be `max_bytes` or `allocatable.bytes`?

Comment (mtojek) [reply x-pack/metricbeat/module/gcp/gke/_meta/fields.yml:131]: In this case - `memory.allocable.bytes`. But this comment applies to all fields you submitted.

Comment (endorama) [reply x-pack/metricbeat/module/gcp/gke/_meta/fields.yml:131]: silly question, it should be `allocatable.bytes` :) Will update all occurrences

**Review by kaiyan-sheng — COMMENTED** [review]
I see this PR is still in draft, just want to bring https://github.com/elastic/ecs/pull/1441 to your attention. As a part of the inventory schema work, we are planning to add some additional container metrics. It would be good if you can adopt these fields in this new metricset 🙂

**Comment by endorama** [comment]
To keep consistency with the GCP documentation and implementation, I'll rename `gke` to `kubernetes`. It's the name used in the [GCP docs](https://cloud.google.com/monitoring/api/metrics_kubernetes) so I think the consistency is a must here.

**Comment by endorama** [comment]
Found a bug in #26960, fix split in #27286

**Comment by endorama** [comment]
Added GKE overview dashboard. Preview:
Screenshot 2021-08-11 at 12-14-55  Metricbeat GCP  GKE Overview 2nd - Elastic

**Comment by endorama** [comment]
https://github.com/elastic/beats/pull/27416 has been merged

Rebasing onto `master` and triggering a new CI check.

**Review by kaiyan-sheng — APPROVED** [review]
LGTM! Feel free to ship the dashboard separately.

**Comment by kaiyan-sheng** [comment]
When shipping a dashboard, please also add a screenshot into this [folder](https://github.com/elastic/beats/tree/master/metricbeat/docs/images) under Metricbeat. Also reference the dashboard screenshot in the [documentation](https://github.com/elastic/beats/blob/master/x-pack/metricbeat/module/gcp/_meta/docs.asciidoc).

**Comment by endorama** [comment]
I removed the dashboard... Now waiting on the CI

**Pull request merged**